### PR TITLE
{CHAT-2, CHAT-6] Registration

### PR DESCRIPTION
**Pull request by vgechev**
- Added logic for registering with phone number
- Added initial implementation for verifying SMS codes
- Added new table UserLoginCodes and migrations for it
- Added logic for generating new login codes and attaching them to the users
- Moved a few files under the correct directory (DTOs moved to DTO folder instead of Models folder) and deleted the later empty folders.
- Added a new validation attribute for the phone number using regex. (example of expected format: +45123456789)
- Added a few validation here and there
- Improved the Result patter a bit by having default parameters
- Formatted a few documents

### CODE REVIEW

**Review by NikolaySlavov-96 — APPROVED** [review]

File: ChatAPI.Application/UseCases/Abstractions/IUserRepository.cs [hunk 1]
```diff
@@ -1,12 +1,13 @@
...
-	public interface IUserRepository
-	{	
-		Task AddUser(UserRegisterDTO user);
-		Task<User?> GetUserByUserId(int userId);
-		Task<User?> GetUserByEmail(string email, bool trackEntity = true);
+    public interface IUserRepository
+	{
+		void AddUser(UserRegisterDTO dto);
+		Task<User?> GetUserByEmailAsNoTracking(string email);
+		Task<UserDevice?> GetUserDeviceByDeviceIdAndPhoneWithUserIncluded(string phone, string deviceId);
```
Comment (NaumVitanov) [inline ChatAPI.Application/UseCases/Abstractions/IUserRepository.cs]: I suggest renaming this to "GetUserDeviceUserIncluded" - the parameters phone and deviceId show how we are going to get the info. Also, right now there is no other method "GetUserDevice...", so I do not think it is necessary to specify how we are getting it.

File: ChatAPI.Domain/Entities/User.cs [hunk 2]
```diff
@@ -2,26 +2,20 @@
 
 namespace ChatAPI.Domain.Entities
 {
-	public class User
+	public class User(string name, string phone)
```
Comment (NaumVitanov) [inline ChatAPI.Domain/Entities/User.cs:5]: Why is this not handled?

File: ChatAPI.Domain/Entities/User.cs [hunk 3]
```diff
@@ -2,26 +2,20 @@
...
-
 		public int Id { get; set; }
 
 		[Required]
-		public required string Name { get; set; }
+		public string Name { get; set; } = name;
 
 		[Required]
-		public required string Phone { get; set; }
+		public string Phone { get; set; } = phone;
```
Comment (NaumVitanov) [inline ChatAPI.Domain/Entities/User.cs:13]: Why did you choose this approach? I mean removing the required key word and using a primary constructor.

File: ChatAPI.Domain/Entities/UserDevice.cs [hunk 4]
```diff
@@ -13,13 +13,21 @@ public UserDevice(int userId, string deviceId, bool isVerified)
 			CreatedOn = DateTime.UtcNow;
 		}
 
-		public int Id { get; set; }
+        public UserDevice(User user, string deviceId)
```
Comment (NaumVitanov) [inline ChatAPI.Domain/Entities/UserDevice.cs:16]: Why is the github action warning not handled?

File: ChatAPI.Application/UseCases/Implementations/UserService.cs [hunk 5]
```diff
@@ -1,26 +1,52 @@
...
+				return Result<SecretLoginCodeDTO>.Failure("Invalid phone and device id combination.");
+
+			// The following is a custom implementation until we actually implement Twilio
+			if (dto.SmsVerificationCode != "000000")
+				return Result<SecretLoginCodeDTO>.Failure("Wrong sms verification code.");
+
+			Guid secretLoginCode = Guid.NewGuid();
+
+			userDevice.IsVerified = true;
+			userDevice.User.UserLoginCode = new(userDevice.UserId, secretLoginCode);
```
Comment (NaumVitanov) [inline ChatAPI.Application/UseCases/Implementations/UserService.cs]: I would like for you to help me understand this better over a call.

File: ChatAPI.Domain/Validations/Attributes/PhoneAttribute.cs [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+		string phoneNumber = value.ToString()!;
+
+		return IsValidPhoneNumber(phoneNumber)
+			? ValidationResult.Success 
+			: new ValidationResult("The phone number is not valid.");
+	}
+
+	private static bool IsValidPhoneNumber(string phone) =>
+		Regex.IsMatch(phone, pattern);
```
Comment (NaumVitanov) [inline ChatAPI.Domain/Validations/Attributes/PhoneAttribute.cs:22]: Can we use the already crated numbers only attribute and just add the "+" sign in front?

File: ChatAPI.Persistence/Repositories/UserRepository.cs [hunk 7]
```diff
@@ -1,34 +1,42 @@
...
+				.Where(d => d.DeviceId == deviceId && d.User.Phone == phone)
+				.FirstOrDefaultAsync();
 
-			if (!trackEntity)
-				query = query.AsNoTracking();
+		public Task SaveChangesAsync() => dbContext.SaveChangesAsync();
 
-			return query.FirstOrDefaultAsync(x => x.Name == email);
-		}
+		public Task<User?> GetUserByEmailAsNoTracking(string email) =>
```
Comment (NaumVitanov) [inline ChatAPI.Persistence/Repositories/UserRepository.cs:39]: Do I understand correctly that this is forgotten?

File: ChatAPI.WebAPI/Controllers/AuthorizationController.cs [hunk 8]
```diff
@@ -19,18 +18,31 @@ public class AuthorizationController(IUserService userService, ITokenService tok
...
 		{
 			Result<TokensDTO> result = await userService.Login(payload);
 
-			if (!result.IsSuccess)
-				return StatusCode(result.StatusCode, result.Error);
+			if (result.IsSuccess)
+				return result.Data;
 
-			return result.Data;
+			return StatusCode(result.StatusCode, result.Error);
```
Comment (NaumVitanov) [inline ChatAPI.WebAPI/Controllers/AuthorizationController.cs:42]: I like the switch here, makes it easier to read.

File: ChatAPI.WebAPI/Controllers/AuthorizationController.cs [hunk 9]
```diff
@@ -19,18 +18,31 @@ public class AuthorizationController(IUserService userService, ITokenService tok
...
 		}
 
+		[Obsolete]
```
Comment (NaumVitanov) [inline ChatAPI.WebAPI/Controllers/AuthorizationController.cs:45]: I like this.

**Review by NaumVitanov — COMMENTED** [review]
I really like the change from passing a parameter to show if we are tracking the entity, to separate methods. Why are the github actions not handled? I do not like the current practice with DTO naming of the models. Are you open to discussion on the topic? How is it that you are using two different accounts to make changes? 
image

File: ChatAPI.Domain/Validations/Attributes/PhoneAttribute.cs (hunk 6 above)
Comment (vgechev) [reply ChatAPI.Domain/Validations/Attributes/PhoneAttribute.cs:22]: no.

**Review by NaumVitanov — APPROVED** [review]
Nice

**Pull request merged**